tests/pass/float.rs

```diff
@@ -12,11 +12,37 @@ use std::fmt::{Debug, Display, LowerHex};
 use std::hint::black_box;
 use std::{f32, f64};
 
+// accept up to 64ULP (16ULP for host floats and 16ULP for artificial error and 32 for any rounding errors)
+const ALLOWED_ULP_DIFF: i128 = 64;
+
 macro_rules! assert_approx_eq {
-    ($a:expr, $b:expr) => {{
+    ($a:expr, $b:expr, $ulp:expr) => {{
         let (a, b) = (&$a, &$b);
-        assert!((*a - *b).abs() < 1.0e-6, "{} is not approximately equal to {}", *a, *b);
+        let a_signum = a.signum();
+        let b_signum = b.signum();
+
+        if a_signum != b_signum {
+            // checking -0.0 == +0.0
+            assert!(a == b, "{:?} is not approximately equal to {:?}: signs differ", *a, *b);
+        }
+        let a_bits = a.to_bits() as i128;
+        let b_bits = b.to_bits() as i128;
```

Comment:
Please add comments explaining what happens here to people not familiar with the binary representation of floats.